StreamChatClient_v3/Controllers/EntityDatabaseObserver.swift

```diff
@@ -0,0 +1,119 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+/// This enum describes the changes to a certain item when observing it.
+public enum EntityChange<Item> {
+    /// The item was created or the recent changes to it make it match the predicate of the observer.
+    case create(_ item: Item)
+    
+    /// The item was updated.
+    case update(_ item: Item)
+    
+    /// The item was deleted or it no longer matches the predicate of the observer.
+    case remove(_ item: Item)
+}
+
+extension EntityChange: Equatable where Item: Equatable {}
+
+extension EntityChange {
+    /// Create a `EnitityChange` value from the provided `ListChange`. It simply transforms `ListChange` to `EntityChange`
+    /// be removing the cases that don't make sense for a single entity change, and using a better naming.
+    init(listChange: ListChange<Item>) {
+        switch listChange {
+        case .insert(let item, index: _):
+            self = .create(item)
+        case let .move(item, _, _), .update(let item, index: _):
+            self = .update(item)
+        case .remove(let item, index: _):
+            self = .remove(item)
+        }
+    }
+}
+
+/// Observes changes of a single entity specified using an `NSFetchRequest`in the provided `NSManagedObjectContext`.
+class EntityDatabaseObserver<Item, DTO: NSManagedObject> {
+    /// The observed item. `nil` of no item matches the predicate or the item was deleted.
+    @Cached var item: Item?
+    
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerDidChangeContent`
+    /// on its delegate.
+    var onChange: ((EntityChange<Item>) -> Void)? {
+        didSet {
+            changeAggregator.onChange = { [unowned self] listChanges in
+                log.assert(listChanges.count <= 1, "EntityDatabaseObserver predicate shouldn't produce more than one change")
```

Comment:
Minor formatting suggestion:
```swift
guard let entityChange = listChanges.first.map(EntityChange.init) else {
 log.assert(true, "EntityDatabaseObserver predicate shouldn't produce more than one change")
 return
}
```

Reply:
But your version changes to behavior. The assertion won't be triggered when there's more than 1 result. Instead, it will be triggered when there's no result, which is a valid case.